File: web/src/state.ts
/**
 * Client-side mirror of server state.
 *
 * The server sends two top-level updates:
 *   - Snapshot  ``{lobby, you, stats, phase, game?}`` — replace-on-receipt.
 *   - Event     individual log entries; append to ``log``.
 *
 * Board layout is fetched once from ``/api/board`` and cached in ``board``.
 */

export interface LobbyPlayer {
  username: string;
  color: string;
  connected: boolean;
}

export interface LobbyState {
  players: LobbyPlayer[];
  mode: "fast" | "slow";
  started: boolean;
}

export interface PlayerStats {
  username: string;
  games_played: number;
  wins: number;
  jewels_stolen: number;
  coins_stolen: number;
  combat_wins: number;
  combat_losses: number;
  racked_count: number;
  imprisoned_count: number;
  tower_cards_gained: number;
  raven_cards_triggered: number;
  doubles_rolled: number;
  total_steps_taken: number;
}

// --- Game ------------------------------------------------------------------

export interface Card {
  id: string;
  kind: string;
  name: string;
  category?: string | null;
  value?: number;
  defender_only?: boolean;
  effect_key?: string | null;
  params?: Record<string, unknown>;
}

export interface GamePlayer {
  username: string;
  color: string;
  position: string;
  hand: Card[];        // own hand (empty array for opponents)
  hand_size: number;
  has_coin: boolean;
  /** Jewels still in your pockets — stealable in a fight. */
  jewels: string[];
  /** Jewels carried out through the Cradle Tower. Safe, and the only ones that
   *  score. */
  banked_jewels: string[];
  accredited: boolean;
  trying_accreditation: boolean;
  status: string;
  status_turns_remaining: number;
  miss_next_turn: boolean;
  connected: boolean;
}

export interface PendingMove {
  steps: number;
  destinations?: Record<string, string[]>;
  remaining_steps?: number;
  has_destinations?: boolean;
  /** Destinations whose only route crosses a manned Yeoman Warder post;
   *  committing to one spends a Disguise from hand. */
  requires_disguise?: string[];
  /** True when the roller is choosing where the split-7 *target* moves. */
  is_for_target?: boolean;
  /** The target player's username when is_for_target is set. */
  target_for_split?: string;
}

export interface Combat {
  attacker: string;
  defender: string;
  space_id: string;
  attacker_cards: Card[];
  defender_cards: Card[];
  attacker_cards_count?: number;
  defender_cards_count?: number;
  attacker_committed: boolean;
  defender_committed: boolean;
  phase: "attacker_selecting" | "defender_selecting" | "defender_specials" | "revealed" | "resolved";
  sanctuary_cancelled: boolean;
  mass_accretor_played: boolean;
  winner: string | null;
  resolved_events: string[];
}

export interface PendingRavenEffect {
  effect_key: string;
  card_id: string;
  params: Record<string, unknown>;
  drawer: string;
}

export interface RavenNotice {
  card_id: string;
  effect_key: string;
  drawer: string;
  params: Record<string, unknown>;
  /** Turned face-up by the drawer. Server state, so the table flips together
   *  — and the card's effect doesn't fire until it's true. */
  revealed: boolean;
}

export interface ConfinementNotice {
  username: string;
  status: "IMPRISONED" | "TORTURED" | "RACKED";
  space_id: string;
  turns: number;
  /** Why they're in there, for the banner copy; "" if the route is unknown. */
  cause: string;
}

export interface PendingCardChange {
  /** "change" = discard one, draw the top of the tower deck (ww41/58/69).
   *  "swap"   = give a chosen card to a chosen opponent, get a random one back. */
  kind: "change" | "swap";
  space_id: string;
  /** Opponents eligible for a swap; empty for a plain change. */
  candidates: string[];
}

export interface PendingSplitSeven {
  total: number;
  source: "seven" | "binary_disruption";
  /** username → the leg sizes that would actually move them. A player with no
   *  entry here can't be given any of the roll. */
  movable_targets: Record<string, number[]>;
  /** The leg sizes the roller may hand over: any split for a natural seven,
   *  but only the two dice as thrown for a Binary Disruption. */
  allowed_legs?: number[];
}

export interface TurnContext {
  roll: number[];
  consecutive_doubles: number;
  extra_turns_queued: number;
  visited_this_turn: string[];
  cards_played_this_turn: string[];
  pending_move: PendingMove | null;
  pending_raven: PendingRavenEffect | null;
  pending_jewel: PendingJewelAttempt | null;
  pending_split: PendingSplitSeven | null;
  pending_card_change: PendingCardChange | null;
}

export interface PendingJewelAttempt {
  jewel_id: string;
  space_id: string;
  /** Whose attempt it is — not always whose turn it is, since a split 7 can
   *  shove an opponent onto a jewel. */
  player: string;
  source: "landing" | "raven_view";
}

export interface GameSnapshot {
  mode: "fast" | "slow";
  phase: string;
  players: GamePlayer[];
  turn_order: string[];
  current_turn_index: number;
  turn: TurnContext;
  jewels_available: Record<string, string>;
  loose_jewels: Record<string, string[]>;
  /** Coins still sitting in the Devereux Tower. */
  coins_available: number;
  /** Size of the pile the game started with — one coin per player. */
  coins_total: number;
  warders: { id: string; location: string }[];
  combat: Combat | null;
  tower_draw_count: number;
  tower_discard_count: number;
  raven_draw_count: number;
  raven_discard_count: number;
  winner: string | null;
  /** Per-player tallies for the results screen, filled in when the game ends. */
  final_stats?: Record<string, Record<string, number>>;
  firecrackers_affected?: string[];
  active_raven_notice?: RavenNotice | null;
  /** Red banner shown to the whole table when somebody is locked up. Only the
   *  player named on it may dismiss it. */
  active_confinement_notice?: ConfinementNotice | null;
  seed: number;
}

// --- Board -----------------------------------------------------------------

export interface BoardSpace {
  id: string;
  label: string;
  region: string;
  kind: string;
  coords?: [number, number];
  coords_region?: [number, number][];
  neighbors: string[];
  wall_walk_order?: number | null;
  action?: { key: string; params?: Record<string, unknown> } | null;
}

export interface BoardData {
  spaces: BoardSpace[];
  // NOTE: /api/board serves the raw board.json, so these arrive with the JSON
  // key names (from_space / to_space), not the Pydantic field aliases.
  slides: {
    id?: string;
    from_space: string;
    to_space: string;
    bidirectional?: boolean;
    path_coords?: [number, number][];
  }[];
  // NOTE: like `slides`, these arrive with the raw JSON key names.
  traversal_edges: {
    id?: string;
    from_space: string;
    to_space: string;
    item?: string;
    description?: string;
    /** Part of the board itself (the secret passage) rather than card-gated. */
    built_in?: boolean;
    direction?: "bidirectional" | "forward" | "backward";
    requires_card?: string | null;
  }[];
  start_space: string;
  escape_space: string;
  queens_house_space: string;
  devereux_space: string;
  rack_space: string;
  bloody_tower_space: string;
  beauchamp_tower_space: string;
  bowyer_tower_space: string;
  chapel_royal_space: string;
  chapel_st_john_space: string;
  hospital_space: string;
  museum_space: string;
  royal_armouries_space: string;
  shop_space: string;
  barracks_space: string;
  warder_posts: { id: string; space_id: string }[];
  raven_deck_space: string;
  metallicity_destination_ids: string[];
  bench_space_ids: string[];
  display_regions: { id: string; label: string; coords_region: [number, number][] }[];
  jewel_display_offset: { x: number; y: number };
  rules: Record<string, unknown>;
}

// --- Top-level client state -----------------------------------------------

export interface LogEntry {
  kind: string;
  payload: Record<string, any>;
  ts: number;
}

export interface ClientState {
  connected: boolean;
  you: string | null;
  lobby: LobbyState | null;
  stats: PlayerStats | null;
  lastError: string | null;
  chat: Array<{ from: string; text: string }>;

  board: BoardData | null;   // fetched once
  game: GameSnapshot | null; // null while in the lobby
  log: LogEntry[];
}

export function initialState(): ClientState {
  return {
    connected: false,
    you: null,
    lobby: null,
    stats: null,
    lastError: null,
    chat: [],
    board: null,
    game: null,
    log: [],
  };
}

export function currentTurnUsername(g: GameSnapshot | null): string | null {
  if (!g || !g.turn_order.length) return null;
  return g.turn_order[g.current_turn_index] ?? null;
}

export function playerByName(g: GameSnapshot | null, name: string | null): GamePlayer | null {
  if (!g || !name) return null;
  return g.players.find((p) => p.username === name) ?? null;
}

File: server/game/state.py
"""Pydantic v2 state models for the Outrage engine.

All state lives in a single :class:`GameState`. Sub-phase machinery (combat,
turn context, pending raven/jewel prompts) is expressed as nested models so a
snapshot of ``GameState`` is sufficient to resume a game.
"""
from __future__ import annotations

from enum import Enum
from typing import Any, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field

from .cards import Card


# ---------- enums ----------------------------------------------------------


class Phase(str, Enum):
    LOBBY = "LOBBY"
    TURN_START = "TURN_START"
    PRE_ROLL = "PRE_ROLL"
    ROLLING = "ROLLING"
    MOVING = "MOVING"
    CHOOSING_PATH = "CHOOSING_PATH"
    COMBAT = "COMBAT"
    JEWEL_ATTEMPT = "JEWEL_ATTEMPT"
    CARD_CHANGE = "CARD_CHANGE"
    ACCREDITATION_ATTEMPT = "ACCREDITATION_ATTEMPT"
    RAVEN_EFFECT = "RAVEN_EFFECT"
    SPLIT_SEVEN_ASSIGN = "SPLIT_SEVEN_ASSIGN"
    TURN_END = "TURN_END"
    GAME_OVER = "GAME_OVER"


class Status(str, Enum):
    NORMAL = "NORMAL"
    HOSPITAL = "HOSPITAL"
    RACKED = "RACKED"
    IMPRISONED = "IMPRISONED"
    TORTURED = "TORTURED"


#: Statuses in which a player's piece is locked in place: they take no turn of
#: their own and nobody else may move them either. The Hospital is deliberately
#: not one of these — it costs a turn but the piece is free.
CONFINED_STATUSES: tuple["Status", ...] = (Status.IMPRISONED, Status.TORTURED, Status.RACKED)


CombatPhase = Literal[
    "attacker_selecting",
    "defender_selecting",
    "defender_specials",
    "revealed",
    "resolved",
]


JewelId = Literal[
    "sword",
    "sceptre",
    "orb",
    "crown_prince_of_wales",
    "crown_st_edward",
]


# ---------- per-player ------------------------------------------------------


class PlayerState(BaseModel):
    model_config = ConfigDict(extra="forbid")

    username: str
    color: str
    position: str
    hand: list[Card] = Field(default_factory=list)
    has_coin: bool = False
    jewels: list[JewelId] = Field(default_factory=list)
    accredited: bool = False
    trying_accreditation: bool = False
    status: Status = Status.NORMAL
    status_turns_remaining: int = 0
    miss_next_turn: bool = False
    # Extra turns owed to this player but not yet taken. Set when something
    # grants an extra turn to somebody who is *not* the acting player — a
    # split-7 leg that shoves an opponent onto the extra-turn square, say.
    # ``_intent_end_turn`` moves it into ``turn.extra_turns_queued`` when play
    # reaches them, so they get the bonus on their own turn, not the roller's.
    extra_turns_pending: int = 0
    connected: bool = True
    #: Jewels carried out through the Cradle Tower and stashed in the hideout.
    #: Safe forever — no fight or pickpocket can touch them — and the only
    #: jewels that score. ``jewels`` above is what is still in your pockets and
    #: still at risk. Using the exit moves the one list into the other.
    banked_jewels: list[JewelId] = Field(default_factory=list)

    # --- derived ----------------------------------------------------------

    @property
    def confined(self) -> bool:
        """Locked up: in the Bloody/Beauchamp/Bowyer Tower or on the Rack.

        A confined piece cannot be moved by anything — not a split 7, not a
        Lasso, not a Sanctuary played by its own owner. Serving the sentence is
        the only way out (bar a Pardon or a Confession).
        """
        return self.status in CONFINED_STATUSES

    # --- convenience mutators ---------------------------------------------

    def add_card(self, card: Card) -> None:
        self.hand.append(card)

    def remove_card(self, card_id: str) -> Optional[Card]:
        for i, c in enumerate(self.hand):
            if c.id == card_id:
                return self.hand.pop(i)
        return None


# ---------- warders ---------------------------------------------------------


class Warder(BaseModel):
    model_config = ConfigDict(extra="forbid")
    id: str
    location: str  # barracks id or post space id


# ---------- combat ---------------------------------------------------------


class Combat(BaseModel):
    model_config = ConfigDict(extra="forbid")

    attacker: str
    defender: str
    space_id: str
    attacker_cards: list[Card] = Field(default_factory=list)
    defender_cards: list[Card] = Field(default_factory=list)
    attacker_committed: bool = False
    defender_committed: bool = False
    phase: CombatPhase = "attacker_selecting"
    # Transient effects:
    sanctuary_cancelled: bool = False
    mass_accretor_played: bool = False
    winner: Optional[str] = None  # set after reveal
    resolved_events: list[str] = Field(default_factory=list)


# ---------- turn context ---------------------------------------------------


class PendingRavenEffect(BaseModel):
    """Tracks a raven card awaiting player input."""

    model_config = ConfigDict(extra="forbid")

    effect_key: str
    card_id: str
    params: dict[str, Any] = Field(default_factory=dict)
    drawer: str


class PendingJewelAttempt(BaseModel):
    """Player must choose which burglary cards to play before we roll."""

    model_config = ConfigDict(extra="forbid")

    jewel_id: JewelId
    space_id: str
    source: Literal["landing", "raven_view"] = "landing"


class PendingCardChange(BaseModel):
    """A "Change a card" square is waiting for the player to pick a discard.

    ``kind`` distinguishes the two prompts that share this shape: the
    wall-walk *change* squares (discard one, draw the top of the tower deck)
    and ww75's *swap* (give one card to a chosen opponent, receive a random one
    back). For a swap, ``candidates`` lists the opponents who can be picked.
    """

    model_config = ConfigDict(extra="forbid")

    kind: Literal["change", "swap"] = "change"
    space_id: str
    candidates: list[str] = Field(default_factory=list)


class PendingSplitSeven(BaseModel):
    model_config = ConfigDict(extra="forbid")

    total: int = 7  # usually 7, but Binary Disruption can reuse this struct
    source: Literal["seven", "binary_disruption"] = "seven"
    # username -> the leg sizes that would actually move them somewhere. A
    # player who is boxed in (an un-accredited piece parked on Queen's House,
    # say) can't be given any of the roll, so the client must not offer them —
    # and if nobody is movable the split never happens at all.
    movable_targets: dict[str, list[int]] = Field(default_factory=dict)


class RavenNotice(BaseModel):
    """Public banner announcing a freshly drawn raven card.

    Held on :class:`GameState` (not :class:`TurnContext`) so reconnecting
    clients see the active notice and any player can dismiss it via the
    ``dismiss_raven_notice`` intent.
    """

    model_config = ConfigDict(extra="forbid")

    card_id: str
    effect_key: str
    drawer: str
    params: dict[str, Any] = Field(default_factory=dict)
    # The card lands face-down. Only the drawer may turn it over, and the
    # effect does not fire until they do — so the table sees the card before
    # anyone's piece moves. Shared state, so everyone flips together.
    revealed: bool = False


class ConfinementNotice(BaseModel):
    """Public banner announcing that somebody has just been locked up.

    The table's counterpart to :class:`RavenNotice`: everyone sees it, so the
    imprisonment reads as an event at the table rather than a quiet status
    change in a side panel. Only the player it happened to may dismiss it.

    Lives on :class:`GameState` so it survives the per-turn ``TurnContext``
    reset and reaches reconnecting clients. Purely informational — it gates no
    phase, so a disconnected victim can't wedge the game by never dismissing.
    """

    model_config = ConfigDict(extra="forbid")

    username: str
    status: Status
    space_id: str
    turns: int = 0
    #: Best-effort description of what put them there, for the banner copy.
    #: Derived from the event log, so an unrecognised route yields "".
    cause: str = ""


class PendingMove(BaseModel):
    """Pre-commit move decision after a roll."""

    model_config = ConfigDict(extra="forbid")

    steps: int
    destinations: dict[str, list[str]] = Field(default_factory=dict)
    # Destinations whose only route runs through an occupied Yeoman Warder post.
    # Reaching one spends a Disguise from hand, so the client labels them and
    # the engine never auto-commits one. Display-only: the charge is worked out
    # from the committed path, not from this list.
    requires_disguise: list[str] = Field(default_factory=list)
    # For split movement: if we've already partially moved, remember what's left.
    remaining_steps: int = 0
    # Segmented moves carry a target (who to stop at) during split-7 second leg.
    forced_stop_space: Optional[str] = None
    # Split-7 continuation: if the self-leg entered ``CHOOSING_PATH``, the
    # target's leg is deferred until the player picks a destination. These
    # fields let ``_intent_choose_move_path`` resume the target's movement.
    split_target: Optional[str] = None
    split_target_destination: Optional[str] = None
    # Target-destination selection: the current player is choosing where the
    # split-7 target moves (not where they themselves move).
    is_for_target: bool = False
    target_for_split: Optional[str] = None
    # target_first: after the roller picks the target's destination, how many
    # steps does the roller still need to take (0 = already done / self_first).
    roller_steps_after_target: int = 0


class DeferredSplitLeg(BaseModel):
    """Half of a split-7 that couldn't run when its turn came round.

    Whichever leg goes first can land somewhere that opens a prompt — a raven
    card, a jewel attempt, a "change a card" square. Running the other leg then
    and there would overwrite that phase and strand whoever owes the answer, so
    it's parked here and picked up once the prompt is closed. Dropping it
    instead would silently cost somebody part of a roll that was legitimately
    made.
    """

    model_config = ConfigDict(extra="forbid")

    kind: Literal["roller", "target"]
    steps: int
    target: Optional[str] = None
    target_destination: Optional[str] = None


class TurnContext(BaseModel):
    model_config = ConfigDict(extra="forbid")

    roll: list[int] = Field(default_factory=list)
    consecutive_doubles: int = 0
    cards_played_this_turn: list[str] = Field(default_factory=list)
    extra_turns_queued: int = 0
    # Every space the current player has stood on this turn, in visit order —
    # used to enforce the "no revisit during a single turn" rule across
    # doubles re-rolls, split-7 legs, and forced / teleport moves.
    visited_this_turn: list[str] = Field(default_factory=list)
    pending_move: Optional[PendingMove] = None
    pending_raven: Optional[PendingRavenEffect] = None
    pending_jewel: Optional[PendingJewelAttempt] = None
    pending_split: Optional[PendingSplitSeven] = None
    pending_card_change: Optional[PendingCardChange] = None
    deferred_split_leg: Optional[DeferredSplitLeg] = None
    # For ``binary_disruption`` and split-7: allow the roller to choose splits
    # on an arbitrary roll; we record the effective total here.
    binary_disruption_armed: bool = False
    # Set when a Disguise card is played pre-roll; allows the player to pass
    # through occupied Yeoman Warder posts for the remainder of this turn.
    disguise_used: bool = False


# ---------- log entries ----------------------------------------------------


class LogEntry(BaseModel):
    model_config = ConfigDict(extra="allow")
    kind: str
    payload: dict[str, Any] = Field(default_factory=dict)


class GameStats(BaseModel):
    """Per-player tallies for a single game, for the end-of-game screen.

    Derived by folding the event log at game end rather than by incrementing
    counters all over the rule engine — the log is already the authoritative
    record of everything that happened, and one fold is far easier to keep
    honest than twenty scattered ``+= 1``s.
    """

    model_config = ConfigDict(extra="forbid")

    turns_taken: int = 0
    steps_taken: int = 0
    doubles_rolled: int = 0
    tower_cards_drawn: int = 0
    raven_cards_drawn: int = 0
    jewel_attempts: int = 0
    jewels_collected: int = 0
    coins_picked_up: int = 0
    fights_won: int = 0
    fights_lost: int = 0
    turns_lost: int = 0
    times_locked_up: int = 0


# ---------- GameState ------------------------------------------------------


class GameState(BaseModel):
    model_config = ConfigDict(extra="forbid")

    mode: Literal["fast", "slow"] = "fast"
    players: list[PlayerState] = Field(default_factory=list)
    turn_order: list[str] = Field(default_factory=list)
    current_turn_index: int = 0
    phase: Phase = Phase.LOBBY

    # Decks
    tower_draw: list[Card] = Field(default_factory=list)
    tower_discard: list[Card] = Field(default_factory=list)
    raven_draw: list[Card] = Field(default_factory=list)
    raven_discard: list[Card] = Field(default_factory=list)

    # Jewels + coins
    jewels_available: dict[JewelId, str] = Field(default_factory=dict)  # jewel -> space_id
    loose_jewels: dict[str, list[JewelId]] = Field(default_factory=dict)  # space_id -> jewels
    # The Devereux Tower holds one coin more than there are players, so exactly
    # one player can always be left short. ``coins_total`` is the size of that
    # pile (set at ``start_game``); ``coins_available`` is how many are still
    # sitting in the tower. Coins handed back — the Rack toll, a fight won by
    # someone already holding one — go back onto the pile, never above the cap.
    coins_total: int = 5
    coins_available: int = 5

    # Warders
    warders: list[Warder] = Field(default_factory=list)

    # Sub-phase
    combat: Optional[Combat] = None
    turn: TurnContext = Field(default_factory=TurnContext)

    # RNG + log
    seed: int = 0
    rng_state: Optional[list] = None  # persisted Rng internal state
    log: list[LogEntry] = Field(default_factory=list)

    # Firecrackers: usernames currently "on notice" — if they end their next
    # turn still inside the White Tower they are sent to the Rack; if their
    # movement takes them outside the White Tower they escape the effect.
    firecrackers_affected: list[str] = Field(default_factory=list)

    # Raven notice: a public modal that lingers until any player dismisses
    # it. Survives the per-turn TurnContext reset so late-arriving clients
    # still see it.
    active_raven_notice: Optional[RavenNotice] = None

    # Confinement notice: the red counterpart to the raven notice. Raised
    # automatically by ``rules.apply`` whenever a player's status crosses into
    # confinement, from whichever route; cleared by the victim, by a later
    # confinement, or by their release.
    active_confinement_notice: Optional[ConfinementNotice] = None

    # Win results
    winner: Optional[str] = None
    # Filled in once, when the game ends. Sent in the snapshot so a player who
    # reconnects to a finished game still sees the full result.
    final_stats: dict[str, GameStats] = Field(default_factory=dict)

    # --- helpers ---------------------------------------------------------

    def player(self, username: str) -> PlayerState:
        for p in self.players:
            if p.username == username:
                return p
        raise KeyError(username)

    def current_player(self) -> PlayerState:
        if not self.turn_order:
            raise RuntimeError("No turn order set")
        return self.player(self.turn_order[self.current_turn_index])

    def player_at(self, space_id: str) -> list[PlayerState]:
        return [p for p in self.players if p.position == space_id]

    def jewel_at_space(self, space_id: str) -> Optional[JewelId]:
        for jid, sid in self.jewels_available.items():
            if sid == space_id:
                return jid
        return None
